### PR TITLE
Improve `isfs` folder creation/modification UX

### PR DESCRIPTION
**Pull request by isc-bsaviano**
IMO our existing command for creating an `isfs` workspace folder have 2 weaknesses:

- It can't be re-used to modify an existing folder. 
- It doesn't allow you to add query parameters. The Studio equivalents can be easily accessed in a menu.

This PR changes the command to allow for the addition of query parameters and for the most of the code to be re-used in a new "modify existing `isfs` folder" command. I also updated the docs to show the new UI.

### CODE REVIEW

**Comment by gjsjohnmurray** [comment]
Based only on reading the diffs I found the new step shown in docs/assets/images/ss-query-params.png hard to understand. Seems like I either set only the first checkbox, or only the second, or neither of those but zero or more of the others.

Did you rule out having a single-choice quickpick to handle ordinary, csp or project, followed by the multi-select one if the choice was o"ordinary"?

Did you consider making it a multi-step procedure, so I can back up to a previous step?

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Looking at this with fresh eyes I see the confusion. I think splitting it up into 2 steps makes sense, but I'm not sure what the call the "ordinary" step that leads to the other query params, or what the prompt for that first step should be.

**Comment by isc-rsingh** [comment]
Can't look at this until tomorrow AM

**Comment by isc-rsingh** [comment]
Since project rules out other choices, that should be the first step (title="Open Project", options=yes/no). Then if project isn't selected, go to CSP (title="Show web application files", options=yes/no), then type or filter.

**Comment by isc-bsaviano** [comment]
I prefer John's 2 step approach if I can get the labels right. I'll keep working on this and update the PR when I think of something.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray @isc-rsingh I just pushed a new commit that split up the query parameter QuickPick and changed the language a little. Let me know what you think.

**Comment by isc-rsingh** [comment]
Much improved

**Review by isc-rsingh — APPROVED** [review]

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray If you don't have any comments I'm going to merge this.

**Comment by gjsjohnmurray** [comment]
@isc-bsaviano I'm busy prepping for a webinar we're giving later today. Should be able to look at PRs tomorrow.

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 1]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+  // Prompt the user for the files to show
+  const filterType = await vscode.window.showQuickPick(
+    [
+      {
+        label: "$(file-code) Show web application files.",
+        detail: "Choose a specific web application, or show all.",
+        value: "csp",
+      },
+      {
+        label: "$(files) Show the contents of a server-side project.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:167]:
```suggestion
        label: "$(files) Contents of a Server-side Project",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 2]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+        detail: "Choose an existing project or create a new one.",
+        value: "project",
+      },
+      {
+        label: "$(list-tree) Show files that pass through custom filters.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:172]:
```suggestion
        label: `$(list-tree) Code Files in ${namespace}`,
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 3]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+  const filterType = await vscode.window.showQuickPick(
+    [
+      {
+        label: "$(file-code) Show web application files.",
+        detail: "Choose a specific web application, or show all.",
+        value: "csp",
+      },
+      {
+        label: "$(files) Show the contents of a server-side project.",
+        detail: "Choose an existing project or create a new one.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:168]:
```suggestion
        detail: "Choose an existing project, or create a new one.",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 4]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+  if (filterType.value == "csp") {
+    // Prompt for a specific web app
+    const cspApps = cspAppsForUri(uri);
+    if (cspApps.length == 0) {
+      vscode.window.showInformationMessage("No web applications are configured in this namespace.", "Dismiss");
+      return;
+    }
+    newPath =
+      (await vscode.window.showQuickPick(cspApps, {
+        title: "Pick a specific web application to show, or press 'Escape' to show all.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:197]:
```suggestion
        placeHolder: "Pick a specific web application to show, or press 'Escape' to show all",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 5]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+        value: "csp",
+      },
+      {
+        label: "$(files) Show the contents of a server-side project.",
+        detail: "Choose an existing project or create a new one.",
+        value: "project",
+      },
+      {
+        label: "$(list-tree) Show files that pass through custom filters.",
+        detail: "Choose the filters to apply.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:173]:
```suggestion
        detail: "Filters can be applied in the next step.",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 6]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+    return;
+  }
+
+  let newParams = "";
+  let newPath = "/";
+  if (filterType.value == "csp") {
+    // Prompt for a specific web app
+    const cspApps = cspAppsForUri(uri);
+    if (cspApps.length == 0) {
+      vscode.window.showInformationMessage("No web applications are configured in this namespace.", "Dismiss");
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:192]:
```suggestion
      vscode.window.showWarningMessage(`No web applications are configured to use namespace ${namespace}.`, "Dismiss");
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 7]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+        {
+          label: "$(references) mapped",
+          detail: "Hide files that are mapped from a non-default database.",
+          picked: oldParams.has("mapped"),
+        },
+      ],
+      {
+        ignoreFocusOut: true,
+        canPickMany: true,
+        title: "",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:236]:
```suggestion
        placeHolder: "Add optional filters",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 8]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+    const otherParams = await vscode.window.showQuickPick(
+      [
+        {
+          label: "$(filter) filter",
+          detail: "Comma-delimited list of search options. (i.e. '*.cls,*.inc,*.mac,*.int')",
+          picked: oldParams.has("filter"),
+        },
+        {
+          label: "$(list-flat) flat",
+          detail: "Show a flat list of files. Does not split packages as folders.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:219]:
```suggestion
          detail: "Show a flat list of files. Do not treat packages as folders.",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 9]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+      return;
+    }
+    newParams = `project=${project}`;
+  } else {
+    // Prompt the user for other query parameters
+    const otherParams = await vscode.window.showQuickPick(
+      [
+        {
+          label: "$(filter) filter",
+          detail: "Comma-delimited list of search options. (i.e. '*.cls,*.inc,*.mac,*.int')",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:214]:
```suggestion
          detail: "Comma-delimited list of search options, e.g. '*.cls,*.inc,*.mac,*.int'",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 10]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+          picked: oldParams.has("flat"),
+        },
+        {
+          label: "$(server-process) generated",
+          detail: "Show generated files as well as non-generated.",
+          picked: oldParams.has("generated"),
+        },
+        {
+          label: "$(references) mapped",
+          detail: "Hide files that are mapped from a non-default database.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:229]:
```suggestion
          detail: `Hide files that are mapped into ${namespace} from another code database.`,
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 11]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+          picked: oldParams.has("filter"),
+        },
+        {
+          label: "$(list-flat) flat",
+          detail: "Show a flat list of files. Does not split packages as folders.",
+          picked: oldParams.has("flat"),
+        },
+        {
+          label: "$(server-process) generated",
+          detail: "Show generated files as well as non-generated.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:224]:
```suggestion
          detail: "Also show files tagged as generated, e.g. by compilation.",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 12]
```diff
@@ -179,3 +147,191 @@ export async function getServerManagerApi(): Promise<any> {
...
+  if (!wsFolderUri) {
+    // Select a workspace folder to modify
+    if (vscode.workspace.workspaceFolders == undefined || vscode.workspace.workspaceFolders.length == 0) {
+      vscode.window.showErrorMessage("No workspace folders are open.", "Dismiss");
+      return;
+    } else if (vscode.workspace.workspaceFolders.length == 1) {
+      wsFolder = vscode.workspace.workspaceFolders[0];
+    } else {
+      wsFolder = await vscode.window.showWorkspaceFolderPick({
+        placeHolder: "Pick the workspace folder modify.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:287]:
```suggestion
        placeHolder: "Pick the workspace folder to modify",
```

**Review by gjsjohnmurray — CHANGES_REQUESTED** [review]
I have made some inline suggestions.

I also think this could be improved:

image

The short lowercase labels feel a bit cryptic. I would prefer:
- Filter
- Flat List
- Generated Code
- Mapped Code

Either your switch statement would need to be amended as well, or you could extend QuickPickItem to add, say, a `value` property to each option and use that in your switch.

I also find it confusing that on this dialog `generated` is about **showing** and `mapped` is about **hiding**. Under the covers both are about showing. How about pre-selecting mapped if it's a new definition or if the existing definition doesn't specify a value? Alternatively make the option labels be "Show Generated" and "Hide Mapped".

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray  `mapped` is hide but `generated` is show because by default StudioOpenDialog shows `mapped` files but hides `generated` ones. I can either switch the names like you suggested, or I can select mapped by default and handle that item separately.

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 13]
```diff
@@ -80,82 +81,49 @@ export async function addServerNamespaceToWorkspace(): Promise<void> {
...
+        label: `$(pencil) Edit Code in ${namespace}`,
+        detail: "Documents opened in this folder will be editable.",
+      },
+      {
+        value: FILESYSTEM_READONLY_SCHEMA,
+        label: `$(lock) View Code in ${namespace}`,
+        detail: "Documents opened in this folder will be read-only.",
+      },
+    ],
+    { title: "Choose the type of access", ignoreFocusOut: true }
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:98]:
```suggestion
    { placeHolder: "Choose the type of access", ignoreFocusOut: true }
```
For consistency with the other steps.

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 14]
```diff
@@ -179,3 +147,192 @@ export async function getServerManagerApi(): Promise<any> {
...
+      switch (otherParamName) {
+        case "filter": {
+          // Prompt for filter
+          const filter = await vscode.window.showInputBox({
+            title: "Enter a filter string.",
+            ignoreFocusOut: true,
+            value: oldParams.get("filter"),
+            placeHolder: "*.cls,*.inc,*.mac,*.int",
+            prompt:
+              "Patterns are comma-delimited and may contain both * (any number of characters) and ? (a single character) as wildcards. To exclude items, prefix the pattern with a single quote.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:256]:
```suggestion
              "Patterns are comma-delimited and may contain both * (zero or more characters) and ? (a single character) as wildcards. To exclude items, prefix the pattern with a single quote.",
```

File: src/commands/addServerNamespaceToWorkspace.ts [hunk 15]
```diff
@@ -179,3 +147,192 @@ export async function getServerManagerApi(): Promise<any> {
...
+  }
+
+  // Prompt the user to modify the uri
+  const newUri = await modifyWsFolderUri(wsFolder.uri);
+  if (!newUri) {
+    return;
+  }
+  // Prompt for name change
+  const newName = await vscode.window.showInputBox({
+    title: "Enter a name for the workspace folder.",
```
Comment (gjsjohnmurray) [inline src/commands/addServerNamespaceToWorkspace.ts:317]:
```suggestion
    title: "Enter a name for the workspace folder",
```

**Review by gjsjohnmurray — CHANGES_REQUESTED** [review]
A couple more tweaks.

Re 'mapped', I don't mind whether you opt to change the label or pre-select the checkbox.

Also, when testing the modify path I noticed that the "what to show" quickpick defaults to the first item even though the URI I'm modifying was created using the third item (i.e. it's a regular one for code in a namespace). I think you could pre-select the right item, but only if you use `createQuickPick` and friends rather than the convenience of `showQuickPick`.

While on the subject of that particular quickpick, how about changing the order so "Code Files in XXX" comes first?

**Comment by gjsjohnmurray** [comment]
I just discovered that if I'm trying to add a web-app-type folder from a server I haven't already connected to, I get told that there are no web apps on the server which use the namespace I chose. This happens because `cspAppsForUri` relies on information in the `cspApps` map which only gets put there by an `await checkConnection(false, uri)` which hasn't yet happened.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Good catch, I will change that to fetch the list from the server

**Comment by gjsjohnmurray** [comment]
When I use the Modify feature on a uri that I had manually added `?system=1` this gets dropped.

Did you consider adding a filter option for including system items? Either way, existing queryparams that the dialog doesn't handle should be preserved IMO.

Also noticed that if I start a modify but don't change anything on the way through I get told "Failed to modify...".

And I suggest we be consistent about capitalization of `Server-Side` (in the Modify command vs `Server-side` (in the quickpick option about projects).

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray I didn't create an entry for `system` because it's undocumented. I can do that though. The only other param is `type`, which I do not want to create an entry for.

**Comment by gjsjohnmurray** [comment]
Latest build no longer seems to update the URI when I use the Modify command. Here's my workspace definition:

```json
{
	"folders": [
		{
			"name": "iris231:USER",
			"uri": "isfs://iris231:user/?mapped=0&system=1"
		}
	],
	"settings": {
		"objectscript.showExplorer": false
	}
}
```

When I get to the checkboxes in the Modify pathway the last two are correctly checked. But if I uncheck either or both of these, OK the quickpick, OK the name (changed or unchanged) the uri in my workspace definition doesn't change.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Sorry about that, my last commit should fix it. I now clear out all known query parameters but leave intact ones that I don't know (like `type`).

**Comment by isc-bsaviano** [comment]
Sorry, that got lost in the shuffle. I'll take a look

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Done

**Review by gjsjohnmurray — APPROVED** [review]
LGTM

**Comment by gjsjohnmurray** [comment]
Too late for this PR I know, but maybe this step would be clearer with an explicit "<all applications>" entry in the quickpick:

image

**Comment by isc-bsaviano** [comment]
That's a good suggestion! Maybe an entry called `All`?

**Pull request merged**